Adding recipe for "pygenn-suite”

Agrim-P777 opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| Rust            | `@conda-forge/help-rust`      |
| Go              | `@conda-forge/help-go`        |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Zulip chat][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://conda-forge.zulipchat.com

All apologies in advance if your recipe PR does not receive prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

### **Recipe for [GeNN (GPU enhanced Neuronal Networks)](https://github.com/genn-team/genn) (CPU & CUDA backend)**

This PR adds `pygenn`, a unified recipe that builds **two variants of PyGeNN** from a single source and package name, distinguished by build string:

**Variants**

- `pygenn=*=cpu*` — CPU-only backend (single-threaded), no CUDA required
- `pygenn=*=cuda*` — CUDA-enabled backend using conda-forge modular CUDA (12.9)

conda selects the correct variant automatically based on whether an NVIDIA GPU driver is detected. Users can override this with `CONDA_OVERRIDE_CUDA`.

A single-package recipe ensures:

- no duplication between CPU and CUDA recipes,
- both variants stay synchronized (version, patches, tests),
- easier maintenance for future updates,
- standard conda variant selection via build string rather than separate package names.


Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x]  If static libraries are linked in, the license of the static library is packaged.
- [x]  Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x]  Build number is 0.
- [x]  A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x]  GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x]  When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

Agrim-P777 committed:
Adding pygenn-cuda package

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/pygenn-cuda/meta.yaml```) and found some lint.

Here's what I've got...


For **recipes/pygenn-cuda/meta.yaml**:

 * ❌ Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [35]
 * ❌ This recipe is using a compiler, which now requires adding a build dependence on `{{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/17325598760](https://github.com/conda-forge/conda-forge-webservices/actions/runs/17325598760). Examine the logs at this URL for more detail.</sub>

Agrim-P777 committed:
Fixed selector spacing, added stdlib c compiler in host, skipped macos builds

Agrim-P777 committed:
updated python pinning format

Agrim-P777 committed:
removed compiler pinning in CBC, removed vs2022 from build

neworderofjamie commented:
I confirm I am willing to be a maintainer of this feedstock

Agrim-P777 committed:
fixed stdlib c issue in build:

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pygenn-cuda/meta.yaml```) and found it was in an excellent condition.

Agrim-P777 committed:
added setuptools to host:

Agrim-P777 committed:
added libffi to host and run

Agrim-P777 committed:
Simplified smoketest for cuda

Agrim-P777 committed:
Added missing_dso_whitelist

Agrim-P777 committed:
reverted the smoketest, updated dso whitelist

Agrim-P777 committed:
fixed nvidia driver stub on CI

Agrim-P777 committed:
added cuda-driver-dev

Agrim-P777 committed:
skip cuda test if no driver found

Agrim-P777 commented:
On Linux, we whitelist `libcuda.so.1` (driver-provided) and on Windows `nvcuda.dll.` PyGeNN also builds and ships its own runtime DLLs (`genn_Release_DLL.dll`, `libffi_Release_DLL.dll`, `genn_single_threaded_cpu_backend_Release_DLL.dll`, `genn_cuda_backend_Release_DLL.dll`), which are bundled with the package and not external dependencies, so they are explicitly whitelisted.

tnowotny commented:
I confirm I am willing to be a maintainer of this feedstock

Agrim-P777 commented:
@conda-forge-admin, please ping conda-forge/help-python-c

All checks are green ✅ and the recipe is ready for review. Thanks!

traversaro commented:
See https://github.com/conda-forge/staged-recipes/pull/30907#issuecomment-3506624313 .

Agrim-P777 committed:
Single pygenn recipe via backend build variant

Agrim-P777 committed:
Package pygenn-suite recipe  parsable by conda-recipe-manager

Agrim-P777 committed:
fixed linter issues

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pygenn-suite/meta.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipes/pygenn-suite/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-recipe-manager`. The recipe can only be automatically migrated to the new v1 format if it is parseable by conda-recipe-manager.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/25883486003](https://github.com/conda-forge/conda-forge-webservices/actions/runs/25883486003). Examine the logs at this URL for more detail.</sub>

carterbox commented on recipes/pygenn-suite/conda_build_config.yaml:
```diff
@@ -0,0 +1,63 @@
+# ----------------------------
```
None of these pinnings are needed. We have channel-wide pinnings for python, compilers, etc. The CUDA packages can be synced by using the cuda-version metapackage.

Please familiarize yourself with these docs: https://github.com/conda-forge/cuda-feedstock/blob/main/recipe/README.md

Agrim-P777 replied:
Thanks for the feedback! This has been addressed — the conda_build_config.yaml has been removed entirely in the updated recipe. The new single-output meta.yaml uses {{ compiler('cuda') }} to drive the CUDA build matrix via conda-forge's global pinning, and cuda-version in host: to keep CUDA sub-packages in sync. The build is skipped for python versions > 311, since GeNN is yet to be running versions higher than 311 reliably (this is included in the meta.yaml no separate conda_build.yml file), also no CUDA 13+ for now, since upstream changes are needed to support it

carterbox commented on recipes/pygenn-suite/meta.yaml:
```diff
@@ -0,0 +1,187 @@
+    This recipe builds the Python bindings (PyGeNN) as two separate outputs:
+      - pygenn-cpu  (CPU-only backend)
+      - pygenn-cuda (CUDA backend)
+
+extra:
+  recipe-maintainers:
```
```suggestion
extra:
  feedstock-name: pygenn
  recipe-maintainers:
```

Multi-output recipes shoudl include feedstock-name so that the bot knows which name to use

carterbox commented on recipes/pygenn-suite/meta.yaml:
```diff
@@ -0,0 +1,187 @@
+    accelerating spiking neural network simulations using GPUs and CPUs.
+    This recipe builds the Python bindings (PyGeNN) as two separate outputs:
+      - pygenn-cpu  (CPU-only backend)
+      - pygenn-cuda (CUDA backend)
```
Are these packages co-installable or variants of the same package?

Agrim-P777 replied:
These are mutually exclusive variants, not co-installable — both installed  into site-packages/pygenn/ and would have conflicted at the file level. This has been addressed in the updated recipe: we now build a single package named pygenn and differentiate the two variants using build strings (cpu_py311_0 vs cuda126_py311_0). conda automatically selects the right variant based on whether an NVIDIA driver is detected via the __cuda virtual package. Users can also override explicitly.

carterbox commented on recipes/pygenn-suite/post-link.sh:
```diff
@@ -0,0 +1,40 @@
+  echo "    # export LD_LIBRARY_PATH=\${CONDA_PREFIX}/lib:\${LD_LIBRARY_PATH:-}"
+  echo
+  echo "--- Option B: Use a system CUDA install (example path) ---"
+  echo "    export CUDA_PATH=/usr/local/cuda-12.x"
```
Why would any user want this? They can choose whatever CTK version they want from conda by installing `cuda-version=X.Y`.

Agrim-P777 replied:
Initially, we thought it might be useful for power users who already have a system CUDA installation and may want to point CUDA_PATH to it.
But on reconsideration, pointing users to a system CUDA install is confusing when everything is managed by conda. The updated post-link.sh and post-link.bat no longer mention a system CUDA option, and now only point users toward the conda-managed CUDA path, including selecting a specific toolkit version via cuda-version=X.Y when needed.

carterbox commented on recipes/pygenn-suite/meta.yaml:
```diff
@@ -0,0 +1,187 @@
+        - $RPATH/genn_Release_DLL.dll                 # GeNN core runtime
+        - $RPATH/libffi_Release_DLL.dll               # Private libffi runtime
+        - $RPATH/genn_single_threaded_cpu_backend_Release_DLL.dll  # CPU backend, bundled
+        - $RPATH/genn_cuda_backend_Release_DLL.dll    # CUDA backend, bundled
```
Are these proprietary binaries that cannot be built from source in this recipe and must be fetched from elsewhere?

Agrim-P777 replied:
No, these are not proprietary binaries — they are all built from source during the python setup.py install step. GeNN's setup.py drives msbuild genn.sln on Windows which compiles genn_Release_DLL.dll, libffi_Release_DLL.dll, and the backend DLLs (genn_single_threaded_cpu_backend_Release_DLL.dll, genn_cuda_backend_Release_DLL.dll) directly from the C++ source in the tarball. They are then placed into pygenn/ and bundled as package_data.
	They appear in missing_dso_whitelist purely because conda-build's post-build DSO audit sees the .pyd extension modules linking against these DLLs but can't find them in the standard library search paths — they live in site-packages/pygenn/ as package_data, not in Library/bin. The whitelist tells conda-build these are intentionally bundled, not missing.
This is separate from libcuda.so.1 / nvcuda.dll, which are legitimately absent at build time and provided by the user's NVIDIA driver at runtime — the more conventional use of missing_dso_whitelist.

carterbox commented on recipes/pygenn-suite/meta.yaml:
```diff
@@ -0,0 +1,187 @@
+        - cuda-cudart-dev {{ cuda_cudart_dev }}
+        - cuda-driver-dev {{ cuda_version }}  # [linux]
+        - cuda-cccl {{ cuda_version }}
+        - libffi
```
```suggestion
      build:
        - {{ stdlib("c") }}  # [linux]
        - {{ compiler('cuda') }}
        - {{ compiler('cxx') }}
        - make  # [linux]
        - python
        - setuptools
        - pybind11
        - psutil
      host:
        - python
        - numpy
        - setuptools
        - psutil
        - pybind11
        - cuda-version {{ cuda_compiler_version }}
        - cuda-cudart-dev
        - cuda-driver-dev  # [linux]
        - cuda-cccl
        - libffi
```
CUDA is treated like any other language toolchain. I suggest you do a GitHub search for feedstocks in the conda-forge org who mention cuda-version or who mention the {{ compiler('cuda') }}.

Agrim-P777 replied:
Thanks for the feedback, these recommendations have been implemented.

carterbox commented on recipes/pygenn-suite/meta.yaml:
```diff
@@ -0,0 +1,187 @@
+        - cuda-libraries {{ cuda_libraries }}
+        - cuda-libraries-dev {{ cuda_libraries_dev }}
+        - cuda-driver-dev {{ cuda_version }}  # [linux]
+        - cuda-cccl {{ cuda_version }}
```
You need the entire cuda-toolkit at runtime? That's fine, but some of these constraints are redundant. You probably only need the cuda compiler template and then one metapackage.

Agrim-P777 replied:
GeNN performs just-in-time compilation of CUDA code at runtime — when a user runs a simulation, GeNN generates and compiles CUDA kernels on the fly using nvcc. This means the full CUDA development stack (headers, stubs, and libraries like cuRAND) must be present in the user's environment at runtime, not just at package build time. cuda-libraries-dev is needed because GeNN's generated code always includes <curand_kernel.h>, cuda-driver-dev for <cuda.h>, and cuda-cccl because cuda_runtime.h pulls it in transitively. The individual constraints are necessary to ensure the right versions of each component are available for runtime compilation.

Agrim-P777 committed:
Update recipes/pygenn-suite/meta.yaml

Multi-output recipes includes the feedstock-name so that the bot knows which name to use

Co-authored-by: Daniel Ching <9604511+carterbox@users.noreply.github.com>

Agrim-P777 committed:
made the changes

Agrim-P777 committed:
kept py311 pinnings

Agrim-P777 committed:
modified requirements.build

Agrim-P777 committed:
fixed _cuda requirements.run

Agrim-P777 committed:
trying different names

Agrim-P777 committed:
added cuda-lib-dev in build

Agrim-P777 committed:
compiler_version none issue

Agrim-P777 committed:
fixed win build

Agrim-P777 committed:
win change

Agrim-P777 committed:
single package name

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/pygenn-suite/meta.yaml```) and found some lint.

Here's what I've got...


For **recipes/pygenn-suite/meta.yaml**:

 * ❌ The recipe is not parsable by any of the known recipe parsers (['conda-forge-tick (the bot)', 'conda-recipe-manager', 'conda-souschef (grayskull)']). Please check the logs for more information and ensure your recipe can be parsed.

For **recipes/pygenn-suite/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-forge-tick (the bot)`. Your recipe  may not receive automatic updates and/or may not be compatible with conda-forge's infrastructure. Please check the logs for more information and ensure your recipe can be parsed.
 * ℹ️ The recipe is not parsable by parser `conda-souschef (grayskull)`. This parser is not currently used by conda-forge, but may be in the future. We are collecting information to see which recipes are compatible with grayskull.
 * ℹ️ The recipe is not parsable by parser `conda-recipe-manager`. The recipe can only be automatically migrated to the new v1 format if it is parseable by conda-recipe-manager.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/25904616564](https://github.com/conda-forge/conda-forge-webservices/actions/runs/25904616564). Examine the logs at this URL for more detail.</sub>

Agrim-P777 committed:
linter issues

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pygenn-suite/meta.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipes/pygenn-suite/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-recipe-manager`. The recipe can only be automatically migrated to the new v1 format if it is parseable by conda-recipe-manager.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/30485437835](https://github.com/conda-forge/conda-forge-webservices/actions/runs/30485437835). Examine the logs at this URL for more detail.</sub>

Agrim-P777 committed:
simplifying build section

Agrim-P777 committed:
libraries in build removed

Agrim-P777 committed:
added to host

Agrim-P777 committed:
removed cuda compiler from build

Agrim-P777 committed:
final recipe

Agrim-P777 commented:
@conda-forge-admin, please ping conda-forge/help-python-c

All checks are green ✅ and the recipe is ready for review. Thanks!
